README.md
# NAME

eksimail - Deliver list of eksisozluk entries

# DESCRIPTION

Download, merge & deliver lists of entries from eksisozluk.com

Install `sendmail` if you are not going to use SendGrid Web API:

    sudo apt-get install sendmail

Install `cpanm` and `carton`

    curl -L http://cpanmin.us | perl - App::cpanminus
    cpanm Carton

This will install dependencies into local/

    carton install

# SYNOPSIS

    carton exec ./eksimail --list=daily --from=you@a.com --to=one@b.com --to=two@c.com

or

    perl -I./local/lib/perl5 ./eksimail --list=daily --from=you@a.com --to=one@b.com --to=two@c.com

- Add `--from` and `--to` addresses with command line arguments. You can add more than one receivers.
- If you don't add a `--from` an email won't be sent.
- If you want to send via SendGrid Web API, provide `--sendgrid-api-key`. Otherwise `sendmail` will be used.
- Adjust politeness delay (for web crawl) with `--sleep`. It's set to 5 seconds by default.

# Arguments

## list

There are two choices.

- `weekly`: Top 20 posts from last week, published by eksisozluk.com. Output is saved at `/tmp/{year}-{week_of_year}.html`.
- `daily`: Most popular entries from yesterday, published by eksisozluk.com. Output is saved at `/tmp/{ymd}.html`.

## sendgrid-api-key

Provide SendGrid Web API key to use SendGrid. Otherwise `sendmail` will be used.

## sleep

Amount of seconds to sleep between each request. This is passed to `WWW::Eksi` as politeness delay argument. Defaults to 5 if not provided.

## from

Email address that email will be sent from. If it's not set, an email won't be sent.

## to

Email address to send email to. This argument can be used more than once (see sample above)

# LICENSE

MIT.


CHANGES
2.33 2020-06-23
  - SendGrid Web API support
  - Some utf8 work
  - Use carton

2.32 2019-08-31
  - Compatible with WWW::Eksi 0.32
  - Daily lists use official DEBE now

2.26 2017-01-22
  - Compatible with WWW::Eksi 0.26
  - Renamed to eksimail (daily lists are back!)

2.24 2017-01-15
  - Compatible with WWW::Eksi 0.24
  - Renamed to eksiweekly (no more daily lists)

2.10 2015-04-27
  - Compatible with WWW::Eksisozluk 0.11
  - Renamed to eksidaily (was eksidebe)

2.01 2014-08-07
  - Homepage related material is removed
  - Favchar fix

0.01 2012-12-05
  - original version; created by h2xs 1.23 with options
    -AX Net::Eksisozluk


eksidebe.pl
#Script to download, merge, deliver yesterday's top50 entries from eksisozluk.
#First Run: 05-Dec-2012

#TODO: download images.
#TODO: make use of calibre command line tools.
#TODO: read mail addresses from the database, do not harcode them.


use DateTime; 
use MIME::Lite;     #Mail
use HTML::Entities;     #To decode html-encoded things like &#252;
use open qw/:std :utf8/;  #To get rid of Turkish character problems
#use warnings;



#Define some variables.

$folder_temp="./";   #folder to work on
$file_in_list="$folder_temp"."inlist";
$file_out_html="$folder_temp"."out.html";
$link_stats="https://eksisozluk.com/istatistik/dunun-en-begenilen-entryleri";
$link_entry="https://eksisozluk.com/entry/";
$link_topic="https://eksisozluk.com/";
$dt   = DateTime->now; 
$searchdate = DateTime->now->subtract(days=>1)->ymd;
$filedate   = DateTime->now->subtract(days=>1)->dmy;
$todaydate  = DateTime->now->dmy;
$datecontrolled="ok";
$searchstring = "\\?a=search\\&searchform.when.from=$searchdate";
$wget="wget --no-check-certificate -q -O";
print "### $dt ###\n"; #LOG


#Write some receivers before running.
#Remember to escape @ with \@.
#Sender to kindle must be approved at amazon.com
$sender_address=""; #will be unique for kindle receivers
$sender_replyto="";
$receivers_mail="";
$receivers_kindle=""; #will be replaced by arrays when database is used (todo)



#Download, open then delete the debe list.

system("$wget $file_in_list $link_stats");
open FILE, "$file_in_list" or die;
@lines = <FILE>;
close FILE or die;
print "Downloaded, "; #LOG



#Use regex to load debe entry details into arrays.
#ith entry of array will show ith entry of debe.

for($i=1,$pointer=0 ; $i<51 && $pointer<@lines ; $pointer++){
  if($lines[$pointer]=~/<span class="caption">(.*)<\/span>/){
    $entries_topic[$i]=$1;
    $entries_topic[$i]=decode_entities($entries_topic[$i]);
    if($lines[$pointer-1]=~/%23(\d+)">/){
          $entries_id[$i]=$1;
    }else{
      print "ERROR: Couldn't get entry id from the debe list for i=$i.\n"
    }
    $i++;
  }
}

#Now to collect entries and create the html.
#We have several arrays about entries starting with "entries_". They are:
#id, topic, numberintopic, datepublished, datemodified, datetoprint, author, body, exist.
#..and more arrays added for reference entry stuff. All the same as above except topic and exist:
#entries_ref_id, entries_ref_numberintopic etc. Topic's the same, hence no entries_ref_topic.

$out="<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">\n".
"<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"tr\"><head>\n".
"<meta http-equiv=\"Content-type\" content=\"text/html; charset=utf-8\" />\n".
"<meta http-equiv=\"Content-language\" content=\"tr\" />\n".
"<style>p.big {line-height:150%;text-align:justify;}</style>".
"<style>p.bigref {line-height:150%;margin-left:0.5cm;text-align:justify;}</style>".
"</head><body>\n\n\n".
"<h2>$filedate'&#252;n en be&#287;enilen entryleri</h2>".
"<br><hr>\n\n\n";

for($i=50;$i>0;$i--){
  system("$wget $folder_temp.entry$i $link_entry$entries_id[$i]");
  open FILE, "$folder_temp.entry$i" or die;@lines = <FILE>;close FILE or die;

  for($j=0;$j<@lines;$j++){
    if($lines[$j]=~/<a href="\/(.*)" itemprop="url">.*<\/a>[^<]/){$entries_topic4link[$i]=$1;} 
    $lines[$j]=~s/<sup class=\"ab\"><([^<]*)(data-query=\")([^>]*)\">\*<\/a><\/sup>/<$1$2$3\">\(* $3\)<\/a>/g;
    $lines[$j]=~s/href="\//target="_blank" href="https:\/\/www.eksisozluk.com\//g;
    $lines[$j]=~s/href="/style="text-decoration:none; color:black" href="/g;
    if($lines[$j]=~/<li id=".*" value="(\d+)"/){$entries_numberintopic[$i]=$1;}
    if($lines[$j]=~/"commentTime">(\d\d)\.(\d\d)\.(\d\d\d\d)(\s\d\d\:\d\d)/){$entries_datepublished[$i]=$1.".".$2.".".$3.$4;
    $datecontrol=$1."-".$2."-".$3; if($datecontrol ne $filedate && $datecontrol ne $todaydate){ $datecontrolled="fail";}
    if($lines[$j]=~/"son g.ncelleme zaman.">(.*)<\/time>/){$entries_datemodified[$i]=$1;}}
    if($lines[$j]=~/data-author="(.*)" data-flags/){$entries_author[$i]=$1;}
    if($lines[$j]=~/commentText">(.*)<\/div>/){$entries_body[$i]=$1;}
  }
  $entries_datetoprint[$i]=$entries_datepublished[$i];
  if($entries_datemodified[$i]){$entries_datetoprint[$i].=" ~ ".$entries_datemodified[$i];}
  
  $entries_exist[$i]=1;
  if ($entries_datepublished[$i]==""){ $entries_exist[$i]=0; }
  
#  control
  
#  print "i:\n$i\n\n";
#  print "entries_id:\n$entries_id[$i]\n\n";
#  print "entries_topic:\n$entries_topic[$i]\n\n";
#  print "topic4link:\n$entries_topic4link[$i]\n\n";
#  print "numberintopic:\n$entries_numberintopic[$i]\n\n";
#  print "datepublished:\n$entries_datepublished[$i]\n\n";
#  print "exists:\n$entries_exist[i]\n\n";
#  print "datecontrolled:\n$datecontrolled\n\n";
#  print "datemodified:\n$entries_datemodified[$i]\n\n";
#  print "author:\n$entries_author[$i]\n\n";
#  print "body:\n$entries_body[$i]\n\n";


  if($entries_exist[$i]){ #this checks whether entry is deleted or not. if deleted, don't print anything.
    $out.=  "<h3>$i. <a href=\"$link_topic$entries_topic4link[$i]\" target=\"blank\" style=\"text-decoration:none; color:black\">$entries_topic[$i]</a></h3><p class=\"big\"><b>$entries_numberintopic[$i]. </b> $entries_body[$i]"
    ."</p><h5><div align=\"right\">(<a href=\"https://www.eksisozluk.com/biri/$entries_author[$i]\" target=\"blank\" style=\"text-decoration:none; color:black\">$entries_author[$i]</a>, <a href=\"https://www.eksisozluk.com/entry/$entries_id[$i]\" "
    ."target=\"blank\" style=\"text-decoration:none; color:black\">$entries_datetoprint[$i]</a>)</div></h5>\n\n";  
  }


  #Find first entry of that day and add a reference to it

  if($entries_exist[$i] && $entries_numberintopic[$i]!=1){ #similarly, if debe entry is deleted, don't print the reference.
    system("$wget $folder_temp.temptopic $link_topic$entries_topic4link[$i]$searchstring");
    open FILE2, "$folder_temp.temptopic" or die;@lines2 = <FILE2>;close FILE2 or die;
    
    for($j2=0;$j2<@lines2&&!$entries_ref_id[$i];$j2++){
      if($lines2[$j2]=~/<li id="li(.*)" value="(\d+)"/){$entries_ref_id[$i]=$1;$entries_ref_numberintopic[$i]=$2;
        if($entries_ref_id[$i]!=$entries_id[$i]){#if debe entry is NOT the first entry of that day
          for($j3=$j2;$j3<$j2+10;$j3++){#get the ref entry details
            $lines2[$j3]=~s/<sup class=\"ab\"><([^<]*)(data-query=\")([^>]*)\">\*<\/a><\/sup>/<$1$2$3\">\(* $3\)<\/a>/g;
            $lines2[$j3]=~s/href="\//target="_blank" href="https:\/\/www.eksisozluk.com\//g;
            $lines2[$j3]=~s/href="/style="text-decoration:none; color:black" href="/g;
            if($lines2[$j3]=~/"commentTime">(\d\d)\.(\d\d)\.(\d\d\d\d)(\s\d\d\:\d\d)/){$entries_ref_datepublished[$i]=$1.".".$2.".".$3.$4;
            if($lines2[$j3]=~/"son g.ncelleme zaman.">(.*)<\/time>/){$entries_ref_datemodified[$i]=$1;}}
            if($lines2[$j3]=~/data-author="(.*)" data-flags/){$entries_ref_author[$i]=$1;}
            if($lines2[$j3]=~/commentText">(.*)<\/div>/){$entries_ref_body[$i]=$1;}  
            $entries_ref_datetoprint[$i]=$entries_ref_datepublished[$i];  
          }
          if($entries_ref_datemodified[$i]){$entries_ref_datetoprint[$i].=" ~ ".$entries_ref_datemodified[$i];}
          
          $out.=  "<p class=\"bigref\"><b>>>$entries_ref_numberintopic[$i]. </b> $entries_ref_body[$i]</p><h5><div align=\"right\">(<a href=\"https://www.eksisozluk.com/biri/$entries_ref_author[$i]\" "
          ."target=\"blank\" style=\"text-decoration:none; color:black\">$entries_ref_author[$i]</a>, <a href=\"https://www.eksisozluk.com/entry/$entries_ref_id[$i]\" target=\"blank\" style=\"text-decoration:none; "
          ."color:black\">$entries_ref_datetoprint[$i]</a>)</div></h5>\n\n";  
  }}}}    if($entries_exist[$i]){ $out.="<hr>\n\n"; } #no <hr> if deleted.
} 

system("mv $file_in_list $folder_temp.entry* $folder_temp.temptopic /tmp");

$out.="<h3>fin.</h3></body>";
open OUT, ">$file_out_html" or die; print OUT $out; close OUT;
print "merged,\n";#LOG


if($datecontrolled eq "ok"){ #this makes sure not to send the day before's top list.

  #Send to kindle readers.

my $msg = MIME::Lite->new(
     From    => "$sender_address",
     To      => "$receivers_kindle",
     'Reply-to' => "$sender_replyto",
     Subject => "$filedate",
     Type    => 'multipart/mixed',
 );

 $msg->attach(
     Type     => 'application/html',
     Path     => "$file_out_html",
     Filename => "$filedate".".html",
     Disposition => 'attachment'

 );

 $msg->send or die;



#Send the created html file to email readers.

my $msg = MIME::Lite->new(
    From    => "$sender_address",
    Bcc      => "$receivers_mail",
    Subject => "$filedate",
    Type    => 'multipart/mixed',
);

$msg->attach(
    Type     => "application/html",
    Path     => "$file_out_html",
    Filename => "$filedate".".html",

);

$msg->send or die;
  print "delivered.\n\n" #log

}else{
  
  print "ERROR: Check the following values:\nEntry ID of 50th:$entries_id[50]\nDate Controlled Value:$datecontrolled\n\n";

}

system("mv $file_out_html /tmp");


address_sample.pm
#address.pm is to be used temporarily. This Will be replaced by a database call.
#Be careful for quotes. If you use " ", then you have to escape @ by \.
#Example:
#to_email_all => "test\@example.com",
#You can also use ' ' with no escape. Example:
#to_email_all => 'test@example.com',

#You can edit all a@a.a addresses accordingly. 
#Just make sure to RENAME the file to address.pm and you'll be ok.

#Make sure to put address.pm to your @INC.

package address;
require Exporter;
use strict;

our @ISA                = qw(Exporter);
our @EXPORT             = qw(getAddress); 

my %Address = ( 

	#Mails will be sent from $from. You need to set some address that you have permission to
	#You'd better set some address from your server. Otherwise your mail will end up in junk.
	from         => 'a@a.a',

	#You will need an actual working address here, otherwise kindle mails may not be delivered.
	reply_to	  => 'a@a.a',

	#Email recipients. All and dev can be different sets.
	to_email_all => 'a@a.a,a@a.a',
	to_email_dev => 'a@a.a',

	#Kindle recipients. All and dev can be different sets.
	to_kindle_all=> 'a@a.a,a@a.a',
	to_kindle_dev=> 'a@a.a',

    			   );

sub getAddress {
 return  %Address;
 }

1;


eksi.pl
package EksiWeekly;
our $VERSION='2.24';

use warnings;
use strict;

use WWW::Eksi 0.24;
use DateTime;
use File::Slurp;
use Getopt::Long;
use MIME::Lite;

=head1 NAME

EksiWeekly - Deliver Eksisozluk GHEBE

=head1 DESCRIPTION

Download, merge & deliver last week's top 20 posts from eksisozluk.com.

Saves merged file at `/tmp/{year}-{week_of_year}.html`.

=head1 SYNOPSIS

    perl eksi.pl --send-email --sleep 10 --from=your@email.com --to=one@subscriber.com --to=two@subscriber.com

 - Add `FROM:` and `TO:` addresses with command line arguments. You can add more than one receivers.
 - If you want to deliver emails, make sure to use `--send-email` (otherwise it will not send).
 - Adjust politeness delay (for web crawl) with `--sleep`. It's set to 5 seconds by default.

=cut

my @to_email;
my $from_email;
my $politeness_delay = 5;
my $send_email       = 0;

GetOptions(
  "to=s@"      => \@to_email,
  "from=s"     => \$from_email,
  "sleep=i"    => \$politeness_delay,
  "send-email" => \$send_email,
);

my $e        = WWW::Eksi->new;
my @ghebe    = $e->ghebe($politeness_delay);

my $dow      = DateTime->now->day_of_week;
my $monday   = DateTime->now->subtract(days=>($dow-1));
my $year     = $monday->year;
my $week     = $monday->week_number;
$week        = "0$week" if $week=~/^\d$/;
my $filename = "$year-$week";
my $filepath = "/tmp/$filename.html";


#Start writing to html.
my $out="<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">\n
<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"tr\"><head>\n
<meta http-equiv=\"Content-type\" content=\"text/html; charset=utf-8\" />\n
<meta http-equiv=\"Content-language\" content=\"tr\" />\n
<style>
p.big {
  line-height:150%;
  text-align:justify;
  margin-left: 0.5cm;
  margin-right: 0.5cm;
}
p.bigref {
  line-height:150%;
  text-align:justify;
  margin-left: 0.5cm;
  margin-right: 0.5cm;
}
img{
  width:90%;
  max-width:250px;
  overflow:hidden;
}
body{
  font: 9pt Verdana, sans-serif;
    background-color: #cccccc;
    color: black;
    height: 100%;
}
a {
    color: #000080;
    text-decoration: none;
}
a:hover {
    background-color: #c0c0c0;
}
</style>
</head><body>\n\n\n
<h2>eksiweekly ${year}-${week}</h2>
<br><hr>\n\n\n";


# Write entries to html
my $i=scalar(@ghebe)+1;
foreach my $entry (reverse @ghebe){
  $i--;
  $out.=  "
  <h3>$i. <a href=\"$entry->{topic_url}\" target=\"blank\" style=\"text-decoration:none; color:black\">
  $entry->{topic_title}</a></h3><p class=\"big\" style=\"text-align:justify;\">$entry->{body_processed}
  </p><h5><div align=\"right\">
  (<a href=\"$entry->{author_url}\" target=\"blank\" style=\"text-decoration:none; color:black\">$entry->{author_name}</a>, <a href=\"$entry->{entry_url}\" target=\"blank\" style=\"text-decoration:none; color:black\">$entry->{time_as_seen}, $entry->{fav_count}&#9734;</a>)</div></h5>\n\n<hr>\n\n";
}


# Finish up html & write out
$out.="</body>";
write_file($filepath,$out);


# Send to email subscribers.
if( $send_email ){
  foreach my $subscriber (@to_email){
    my $msg_mail = MIME::Lite->new(
      From    => $from_email,
      To      => $subscriber,
      Subject => "eksiweekly $filename",
      Type    => 'multipart/mixed',
    );

    $msg_mail->attach(
      Type     => 'application/html',
      Path     => "$filepath",
      Filename => "$filename.html",
    );

    $msg_mail->send or next "Error on sending email to $subscriber";
  }
}
